sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java

```diff
@@ -88,108 +98,250 @@ public static void main(String[] args) throws IOException, InterruptedException
+        // listEventRoutes();
+
+        // TODO: Deleting event route
+        // deleteEventRoute();
     }
 
     /**
      * Delete a twin, and any relationships it might have.
      * @throws IOException If an I/O error is thrown when accessing the starting file.
-     * @throws InterruptedException If the current thread is interrupted while waiting to acquire permits on a semaphore.
+     * @throws InterruptedException If the current thread is interrupted while waiting to acquire latch.
      */
     public static void deleteTwins() throws IOException, InterruptedException {
         System.out.println("DELETE DIGITAL TWINS");
         Map<String, String> twins = FileHelper.loadAllFilesInPath(TwinsPath);
-        final Semaphore deleteTwinsSemaphore = new Semaphore(0);
-        final Semaphore deleteRelationshipsSemaphore = new Semaphore(0);
 
         // Call APIs to clean up any pre-existing resources that might be referenced by this sample. If digital twin does not exist, ignore.
-        twins
-            .forEach((twinId, twinContent) -> {
-                // Call APIs to delete all relationships.
-                client.listRelationships(twinId, BasicRelationship.class)
-                    .doOnComplete(deleteRelationshipsSemaphore::release)
-                    .doOnError(throwable -> {
-                        if (throwable instanceof ErrorResponseException && ((ErrorResponseException) throwable).getResponse().getStatusCode() == HttpStatus.SC_NOT_FOUND) {
-                            deleteRelationshipsSemaphore.release();
-                        } else {
-                            System.err.println("List relationships error: " + throwable);
-                        }
-                    })
-                    .subscribe(
-                        relationship -> client.deleteRelationship(twinId, relationship.getId())
-                            .subscribe(
-                                aVoid -> System.out.println("Found and deleted relationship: " + relationship.getId()),
-                                throwable -> System.err.println("Delete relationship error: " + throwable)
-                            ));
-
-                // Call APIs to delete any incoming relationships.
-                client.listIncomingRelationships(twinId)
-                    .doOnComplete(deleteRelationshipsSemaphore::release)
+        // Once the async API terminates (either successfully, or with an error), the latch count is decremented, or the semaphore is released.
+        for (Map.Entry<String, String> twin : twins.entrySet()) {
+            String twinId = twin.getKey();
+
+            List<BasicRelationship> relationshipList = new ArrayList<>();
+            Semaphore listRelationshipSemaphore = new Semaphore(0);
+            Semaphore deleteRelationshipsSemaphore = new Semaphore(0);
+            CountDownLatch deleteTwinsLatch = new CountDownLatch(1);
+
+            // Call APIs to retrieve all relationships.
```

Comment:
This implementation is very long and verbose. I am trying to see how I could chain the list and delete operations together, I'll put up an update in the next PR.